ast/src/parsed/pattern.rs

```diff
@@ -0,0 +1,432 @@
+use powdr_number::BigInt;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+use std::iter::{empty, once};
+
+use super::{asm::SymbolPath, visitor::Children};
+
+/// When a pattern is specialised, in some cases the algorithm will return
+/// a list of its internal patterns. PatternTuple works as a wrapper that makes these
+/// patterns identifiable and implements the specialisation in tuples following
+/// rustc[1] implementation.
+///
+/// [1] https://doc.rust-lang.org/nightly/nightly-rustc/rustc_pattern_analysis/usefulness/index.html#specialization
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct PatternTuple {
+    pub patterns: Vec<Pattern>,
+}
+
+impl PatternTuple {
+    pub fn is_empty(&self) -> bool {
+        self.patterns.is_empty()
+    }
+
+    /// Check for irrefutable patterns in pattern tuples. Since pattern tuples can contain
+    /// internal patterns, irrefutability only applies in cases of unique patterns (len == 1).
+
+    /// Example:
+    /// If (_, _) was obtained from specialization, it is not true that the pattern is irrefutable
+    /// because having more than one element implies that the evaluation is performed on internal patterns.
+    /// If, on the other hand, the pattern tuple is `(Pattern::Tuple(_,_), )``, irrefutability applies in the standard way.
+    pub fn is_irrefutable(&self) -> bool {
+        if self.patterns.len() == 1 {
+            self.patterns[0].is_irrefutable()
+        } else {
+            false
+        }
+    }
+
+    /// Specialize a pattern tuple based on a "constructor" pattern tuple passed as a parameter.
+    /// Based on https://doc.rust-lang.org/nightly/nightly-rustc/rustc_pattern_analysis/usefulness/index.html#specialization.
+    /// If the constructor is shared with the pattern, specialize the pattern removing
+    /// "one layer" (the constructor) and returning internal patterns.
+    /// Return None if the pattern can't be specialized with the constructor or if the pattern is empty.
```

Comment:
I still don't really understand this routine, but maybe adding the following items to the docstring helps:

We are only interested in the outermost type-constructor of the `constructor` pattern and ignore the fields and the inner patterns.

And maybe also a variant of the following:
```
matches!(c(v_1, .., v_n), p)
<=> specialize(c, p) returns something
    && matches!((v_1, .., v_n), specialize(c, p))
```